File: src/schemas/calendar.ts
import { z } from "zod";
import { datasetEnvelope, utcDateTime } from "./envelope";

export const EventType = z.enum(["economic", "earnings"]);
export const EventImportance = z.enum(["high", "medium", "low"]);

export const CalendarEvent = z
  .object({
    id: z.string().min(1),
    type: EventType,
    title: z.string().min(1),
    country: z.string().nullable(),
    datetime: utcDateTime,
    importance: EventImportance,
    actual: z.number().finite().nullable(),
    forecast: z.number().finite().nullable(),
    previous: z.number().finite().nullable(),
    unit: z.string().nullable(),
    related_assets: z.array(z.string()),
    source: z.string().min(1),
  })
  .strict();

export const CalendarDataset = z
  .object({
    events: z.array(CalendarEvent),
    updated_at: utcDateTime.nullable(),
  })
  .strict();

export const CalendarEnvelope = datasetEnvelope(CalendarDataset);

export type EventType = z.infer<typeof EventType>;
export type EventImportance = z.infer<typeof EventImportance>;
export type CalendarEvent = z.infer<typeof CalendarEvent>;
export type CalendarDataset = z.infer<typeof CalendarDataset>;
export type CalendarEnvelope = z.infer<typeof CalendarEnvelope>;

File: pipeline/schemas/calendar.py
"""Calendar dataset contract (economic calendar + earnings calendar, architecture §1.3 / §3.5).

Filled by the T03 CalendarCollector; this module only defines the contract.
"""

from __future__ import annotations

from typing import Literal

from pydantic import Field

from .envelope import BaseEnvelope, ContractModel, UTCDateTime

EventType = Literal["economic", "earnings"]
EventImportance = Literal["high", "medium", "low"]


class CalendarEvent(ContractModel):
    #: Stable dedupe key (#94): earnings-{SYMBOL}-{date} (shared by FMP + Nasdaq),
    #: econ-fred-{release_id}-{date}, econ-fomc-{date}. The collector dedupes by id —
    #: the same event from two sources can never double-publish.
    id: str = Field(min_length=1, description="stable dedupe key, e.g. econ-fred-10-2026-08-12")
    type: EventType
    title: str = Field(min_length=1)
    country: str | None = None
    datetime: UTCDateTime
    importance: EventImportance = "medium"
    actual: float | None = None
    forecast: float | None = None
    previous: float | None = None
    unit: str | None = None
    related_assets: list[str] = Field(default_factory=list)
    source: str = Field(min_length=1)


class CalendarDataset(ContractModel):
    """calendar.json payload."""

    events: list[CalendarEvent] = Field(default_factory=list)
    updated_at: UTCDateTime | None = None


class CalendarEnvelope(BaseEnvelope):
    payload: CalendarDataset
